File: gigvora-flutter-phoneapp/lib/features/governance/domain/consent_models.dart
class ConsentPolicyInfo {
  ConsentPolicyInfo({
    required this.id,
    required this.code,
    required this.title,
    required this.audience,
    required this.region,
    required this.legalBasis,
    required this.required,
    required this.revocable,
    this.retentionPeriodDays,
    this.description,
  });

  final int id;
  final String code;
  final String title;
  final String audience;
  final String region;
  final String legalBasis;
  final bool required;
  final bool revocable;
  final int? retentionPeriodDays;
  final String? description;

  factory ConsentPolicyInfo.fromJson(Map<String, dynamic> json) {
    int? parseInt(dynamic value) {
      if (value == null) return null;
      if (value is int) return value;
      if (value is num) return value.toInt();
      if (value is String) return int.tryParse(value);
      return null;
    }

    return ConsentPolicyInfo(
      id: parseInt(json['id']) ?? 0,
      code: json['code'] as String? ?? '',
      title: json['title'] as String? ?? 'Untitled policy',
      audience: (json['audience'] as String? ?? 'user').toLowerCase(),
      region: (json['region'] as String? ?? 'global').toLowerCase(),
      legalBasis: json['legalBasis'] as String? ?? 'unknown',
      required: json['required'] as bool? ?? false,
      revocable: json['revocable'] as bool? ?? true,
      retentionPeriodDays: parseInt(json['retentionPeriodDays']),
      description: json['description'] as String?,
    );
  }
}

class UserConsentDecision {
  UserConsentDecision({
    required this.id,
    required this.policyId,
    required this.policyVersionId,
    required this.status,
    this.grantedAt,
    this.withdrawnAt,
    this.source,
  });

  final int id;
  final int policyId;
  final int policyVersionId;
  final String status;
  final DateTime? grantedAt;
  final DateTime? withdrawnAt;
  final String? source;

  factory UserConsentDecision.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic value) {
      if (value is String && value.isNotEmpty) {
        return DateTime.tryParse(value);
      }
      return null;
    }

    int parseInt(dynamic value) {
      if (value is int) return value;
      if (value is num) return value.toInt();
      if (value is String) {
        final parsed = int.tryParse(value);
        if (parsed != null) return parsed;
      }
      return 0;
    }

    return UserConsentDecision(
      id: parseInt(json['id']),
      policyId: parseInt(json['policyId']),
      policyVersionId: parseInt(json['policyVersionId']),
      status: (json['status'] as String? ?? 'withdrawn').toLowerCase(),
      grantedAt: parseDate(json['grantedAt']),
      withdrawnAt: parseDate(json['withdrawnAt']),
      source: json['source'] as String?,
    );
  }
}

class ConsentAuditEventInfo {
  ConsentAuditEventInfo({
    required this.id,
    required this.action,
    this.occurredAt,
    this.actorLabel,
    this.metadata,
    this.version,
    this.versionEffectiveAt,
    this.versionSupersededAt,
    this.consentStatus,
    this.consentGrantedAt,
    this.consentWithdrawnAt,
  });

  final int id;
  final String action;
  final DateTime? occurredAt;
  final String? actorLabel;
  final Map<String, dynamic>? metadata;
  final int? version;
  final DateTime? versionEffectiveAt;
  final DateTime? versionSupersededAt;
  final String? consentStatus;
  final DateTime? consentGrantedAt;
  final DateTime? consentWithdrawnAt;

  static DateTime? _parseDate(dynamic value) {
    if (value == null) return null;
    if (value is DateTime) return value;
    if (value is String && value.isNotEmpty) {
      return DateTime.tryParse(value);
    }
    return null;
  }

  static int? _parseInt(dynamic value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is num) return value.toInt();
    if (value is String) {
      return int.tryParse(value);
    }
    return null;
  }

  factory ConsentAuditEventInfo.fromJson(Map<String, dynamic> json) {
    final actor = json['actor'] as Map<String, dynamic>?;
    final actorType = actor?['type'] as String? ?? 'system';
    final actorId = actor?['id'];
    final actorLabel = actorId != null ? '$actorType • $actorId' : actorType;

    final versionJson = json['policyVersion'] as Map<String, dynamic>?;
    final consentJson = json['userConsent'] as Map<String, dynamic>?;

    return ConsentAuditEventInfo(
      id: _parseInt(json['id']) ?? 0,
      action: json['action'] as String? ?? 'unknown_event',
      occurredAt: _parseDate(json['occurredAt']),
      actorLabel: actorLabel,
      metadata: json['metadata'] is Map<String, dynamic> ? json['metadata'] as Map<String, dynamic> : null,
      version: _parseInt(versionJson?['version']),
      versionEffectiveAt: _parseDate(versionJson?['effectiveAt']),
      versionSupersededAt: _parseDate(versionJson?['supersededAt']),
      consentStatus: consentJson?['status'] as String?,
      consentGrantedAt: _parseDate(consentJson?['grantedAt']),
      consentWithdrawnAt: _parseDate(consentJson?['withdrawnAt']),
    );
  }
}

class ConsentSnapshotEntry {
  ConsentSnapshotEntry({
    required this.policy,
    this.consent,
    this.auditTrail = const [],
  });

  final ConsentPolicyInfo policy;
  final UserConsentDecision? consent;
  final List<ConsentAuditEventInfo> auditTrail;

  factory ConsentSnapshotEntry.fromJson(Map<String, dynamic> json) {
    return ConsentSnapshotEntry(
      policy: ConsentPolicyInfo.fromJson(json['policy'] as Map<String, dynamic>),
      consent: json['consent'] == null
          ? null
          : UserConsentDecision.fromJson(json['consent'] as Map<String, dynamic>),
      auditTrail: (json['auditTrail'] as List<dynamic>? ?? const [])
          .map((entry) => ConsentAuditEventInfo.fromJson(entry as Map<String, dynamic>))
          .toList(),
    );
  }
}

class UserConsentSnapshot {
  UserConsentSnapshot({
    required this.userId,
    required this.entries,
    required this.outstandingRequired,
    required this.fetchedAt,
  });

  final int userId;
  final List<ConsentSnapshotEntry> entries;
  final int outstandingRequired;
  final DateTime fetchedAt;

  factory UserConsentSnapshot.fromJson(Map<String, dynamic> json) {
    final policiesJson = json['policies'] as List<dynamic>? ?? const [];
    return UserConsentSnapshot(
      userId: json['userId'] is int
          ? json['userId'] as int
          : int.tryParse('${json['userId']}') ?? 0,
      entries: policiesJson
          .map((item) => ConsentSnapshotEntry.fromJson(item as Map<String, dynamic>))
          .toList(),
      outstandingRequired: json['outstandingRequired'] is int
          ? json['outstandingRequired'] as int
          : int.tryParse('${json['outstandingRequired']}') ?? 0,
      fetchedAt: DateTime.now(),
    );
  }

  int get grantedCount => entries.where((entry) => entry.consent?.status == 'granted').length;
}

File: gigvora-backend-nodejs/src/models/consentModels.js
import { DataTypes } from 'sequelize';
import sequelize from './sequelizeClient.js';

const dialect = sequelize.getDialect();
const jsonType = ['postgres', 'postgresql'].includes(dialect) ? DataTypes.JSONB : DataTypes.JSON;

export const ConsentPolicy = sequelize.define(
  'ConsentPolicy',
  {
    code: {
      type: DataTypes.STRING(80),
      allowNull: false,
      unique: true,
      validate: {
        is: /^[a-z0-9_.-]+$/i,
        len: [2, 80],
      },
    },
    title: {
      type: DataTypes.STRING(240),
      allowNull: false,
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    audience: {
      type: DataTypes.STRING(80),
      allowNull: false,
      defaultValue: 'user',
      validate: {
        is: /^[a-z0-9_.-]+$/i,
      },
    },
    region: {
      type: DataTypes.STRING(32),
      allowNull: false,
      defaultValue: 'global',
      validate: {
        is: /^[a-z0-9_.-]+$/i,
      },
    },
    legalBasis: {
      type: DataTypes.STRING(80),
      allowNull: false,
    },
    required: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
    },
    revocable: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: true,
    },
    retentionPeriodDays: {
      type: DataTypes.INTEGER,
      allowNull: true,
      validate: {
        min: 1,
        max: 3650,
      },
    },
    activeVersionId: {
      type: DataTypes.INTEGER,
      allowNull: true,
    },
    metadata: {
      type: jsonType,
      allowNull: false,
      defaultValue: {},
    },
    createdBy: {
      type: DataTypes.STRING(120),
      allowNull: true,
    },
    updatedBy: {
      type: DataTypes.STRING(120),
      allowNull: true,
    },
  },
  {
    tableName: 'consent_policies',
    indexes: [
      { fields: ['audience', 'region'] },
      { fields: ['activeVersionId'] },
    ],
  },
);

export const ConsentPolicyVersion = sequelize.define(
  'ConsentPolicyVersion',
  {
    policyId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: ConsentPolicy,
        key: 'id',
      },
      onDelete: 'CASCADE',
      onUpdate: 'CASCADE',
    },
    version: {
      type: DataTypes.INTEGER,
      allowNull: false,
      validate: {
        min: 1,
      },
    },
    documentUrl: {
      type: DataTypes.STRING(1024),
      allowNull: true,
      validate: {
        isUrl(value) {
          if (!value) return;
          try {
            // eslint-disable-next-line no-new
            new URL(value);
          } catch (error) {
            throw new Error('documentUrl must be a valid URL when provided.');
          }
        },
      },
    },
    content: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    summary: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    effectiveAt: {
      type: DataTypes.DATE,
      allowNull: false,
      validate: {
        isDate: true,
      },
    },
    supersededAt: {
      type: DataTypes.DATE,
      allowNull: true,
      validate: {
        isDate: true,
      },
    },
    createdBy: {
      type: DataTypes.STRING(120),
      allowNull: true,
    },
    metadata: {
      type: jsonType,
      allowNull: false,
      defaultValue: {},
    },
  },
  {
    tableName: 'consent_policy_versions',
    indexes: [{ fields: ['policyId', 'effectiveAt'] }],
  },
);

export const UserConsent = sequelize.define(
  'UserConsent',
  {
    userId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
      onDelete: 'CASCADE',
      onUpdate: 'CASCADE',
    },
    policyId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: ConsentPolicy,
        key: 'id',
      },
      onDelete: 'CASCADE',
      onUpdate: 'CASCADE',
    },
    policyVersionId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: ConsentPolicyVersion,
        key: 'id',
      },
      onDelete: 'RESTRICT',
      onUpdate: 'CASCADE',
    },
    status: {
      type: DataTypes.ENUM('granted', 'withdrawn'),
      allowNull: false,
    },
    grantedAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    withdrawnAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    source: {
      type: DataTypes.STRING(80),
      allowNull: false,
      defaultValue: 'self_service',
    },
    ipAddress: {
      type: DataTypes.STRING(120),
      allowNull: true,
    },
    userAgent: {
      type: DataTypes.STRING(500),
      allowNull: true,
    },
    metadata: {
      type: jsonType,
      allowNull: false,
      defaultValue: {},
    },
  },
  {
    tableName: 'user_consents',
    indexes: [
      { fields: ['userId', 'policyId'], unique: true },
      { fields: ['policyId', 'status'] },
      { fields: ['policyVersionId'] },
    ],
  },
);

export const ConsentAuditEvent = sequelize.define(
  'ConsentAuditEvent',
  {
    policyId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: ConsentPolicy,
        key: 'id',
      },
      onDelete: 'CASCADE',
      onUpdate: 'CASCADE',
    },
    policyVersionId: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: ConsentPolicyVersion,
        key: 'id',
      },
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE',
    },
    userConsentId: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: UserConsent,
        key: 'id',
      },
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE',
    },
    actorId: {
      type: DataTypes.STRING(120),
      allowNull: true,
    },
    actorType: {
      type: DataTypes.STRING(80),
      allowNull: false,
      defaultValue: 'system',
    },
    action: {
      type: DataTypes.STRING(80),
      allowNull: false,
    },
    reason: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    metadata: {
      type: jsonType,
      allowNull: false,
      defaultValue: {},
    },
  },
  {
    tableName: 'consent_audit_events',
    indexes: [{ fields: ['policyId', 'createdAt'] }],
  },
);

ConsentPolicy.hasMany(ConsentPolicyVersion, { foreignKey: 'policyId', as: 'versions' });
ConsentPolicyVersion.belongsTo(ConsentPolicy, { foreignKey: 'policyId', as: 'policy' });

ConsentPolicy.hasMany(UserConsent, { foreignKey: 'policyId', as: 'consents' });
UserConsent.belongsTo(ConsentPolicy, { foreignKey: 'policyId', as: 'policy' });

ConsentPolicyVersion.hasMany(UserConsent, { foreignKey: 'policyVersionId', as: 'userConsents' });
UserConsent.belongsTo(ConsentPolicyVersion, { foreignKey: 'policyVersionId', as: 'policyVersion' });

ConsentPolicy.hasMany(ConsentAuditEvent, { foreignKey: 'policyId', as: 'auditEvents' });
ConsentAuditEvent.belongsTo(ConsentPolicy, { foreignKey: 'policyId', as: 'policy' });

ConsentPolicyVersion.hasMany(ConsentAuditEvent, { foreignKey: 'policyVersionId', as: 'auditEvents' });
ConsentAuditEvent.belongsTo(ConsentPolicyVersion, { foreignKey: 'policyVersionId', as: 'policyVersion' });

UserConsent.hasMany(ConsentAuditEvent, { foreignKey: 'userConsentId', as: 'events' });
ConsentAuditEvent.belongsTo(UserConsent, { foreignKey: 'userConsentId', as: 'userConsent' });

ConsentPolicy.prototype.toSummary = function toSummary({ includeVersions = false } = {}) {
  const plain = this.get({ plain: true });
  const summary = {
    id: plain.id,
    code: plain.code,
    title: plain.title,
    description: plain.description,
    audience: plain.audience,
    region: plain.region,
    legalBasis: plain.legalBasis,
    required: Boolean(plain.required),
    revocable: Boolean(plain.revocable),
    retentionPeriodDays: plain.retentionPeriodDays ?? null,
    activeVersionId: plain.activeVersionId ?? null,
    metadata: plain.metadata ?? {},
    createdBy: plain.createdBy ?? null,
    updatedBy: plain.updatedBy ?? null,
    createdAt: plain.createdAt ? new Date(plain.createdAt).toISOString() : null,
    updatedAt: plain.updatedAt ? new Date(plain.updatedAt).toISOString() : null,
  };

  if (includeVersions) {
    summary.versions = (plain.versions ?? []).map((version) => ({
      id: version.id,
      version: version.version,
      documentUrl: version.documentUrl ?? null,
      summary: version.summary ?? null,
      effectiveAt: version.effectiveAt ? new Date(version.effectiveAt).toISOString() : null,
      supersededAt: version.supersededAt ? new Date(version.supersededAt).toISOString() : null,
      metadata: version.metadata ?? {},
    }));
  }

  return summary;
};

UserConsent.prototype.toSnapshot = function toSnapshot() {
  const plain = this.get({ plain: true });
  return {
    id: plain.id,
    policyId: plain.policyId,
    policyVersionId: plain.policyVersionId,
    status: plain.status,
    grantedAt: plain.grantedAt ? new Date(plain.grantedAt).toISOString() : null,
    withdrawnAt: plain.withdrawnAt ? new Date(plain.withdrawnAt).toISOString() : null,
    source: plain.source,
    metadata: plain.metadata ?? {},
  };
};

export function normaliseConsentCode(value) {
  if (!value) return null;
  const trimmed = `${value}`.trim().toLowerCase();
  if (!trimmed) return null;
  return trimmed.replace(/[^a-z0-9_.-]/g, '-');
}

export async function activatePolicyVersion(policy, version, { transaction } = {}) {
  const activeAt = new Date();
  await policy.update({ activeVersionId: version.id }, { transaction });
  await ConsentAuditEvent.create(
    {
      policyId: policy.id,
      policyVersionId: version.id,
      action: 'policy_version_activated',
      metadata: { effectiveAt: version.effectiveAt ?? activeAt.toISOString() },
    },
    { transaction },
  );
}

export async function supersedePolicyVersion(version, { supersededAt = new Date(), actorId, transaction } = {}) {
  await version.update({ supersededAt }, { transaction });
  await ConsentAuditEvent.create(
    {
      policyId: version.policyId,
      policyVersionId: version.id,
      actorId: actorId ?? null,
      actorType: actorId ? 'admin' : 'system',
      action: 'policy_version_superseded',
      metadata: { supersededAt: supersededAt.toISOString() },
    },
    { transaction },
  );
}

export default {
  ConsentPolicy,
  ConsentPolicyVersion,
  UserConsent,
  ConsentAuditEvent,
};
